### PR TITLE
Consistent return values for first and last methods

### PR DESCRIPTION
**Pull request by phalkunz**
Issue: https://github.com/silverstripe/silverstripe-framework/issues/3239

### CODE REVIEW

**Comment by flamerohr** [comment]
Since this is quite a core API change, could we add some unit tests to make sure we're covered for future changes in this?

File: src/ORM/ArrayList.php [hunk 1]
```diff
@@ -302,7 +302,8 @@ public function shift()
      */
     public function first()
     {
-        return reset($this->items);
+        $first = reset($this->items);
+        return $first === false ? null : $first;
```
Comment (tractorcow) [inline src/ORM/ArrayList.php]: What if `$this->items` is `[false]`? :P

Comment (tractorcow) [inline src/ORM/ArrayList.php]: Maybe we should have a conditional on `empty($this->items)` instead?

File: src/ORM/ArrayList.php [hunk 2]
```diff
@@ -516,7 +518,7 @@ public function canFilterBy($by)
     {
         $firstRecord = $this->first();
 
-        if ($firstRecord === false) {
+        if ($firstRecord === null) {
```
Comment (tractorcow) [inline src/ORM/ArrayList.php]: If the first record is false OR null then we can't filter.

**Comment by tractorcow** [comment]
Note: `canFilterBy` should, but doesn't, use the same logic as `extractValue`. `extractValue` supports both array / dataobject, but `canFilterBy` only supports array.

File: src/ORM/ArrayList.php (hunk 1 above)
Comment (phalkunz) [inline src/ORM/ArrayList.php]: Ah, true. 👍

File: src/ORM/ArrayList.php [hunk 3]
```diff
@@ -516,7 +524,7 @@ public function canFilterBy($by)
     {
         $firstRecord = $this->first();
 
-        if ($firstRecord === false) {
+        if(empty($this->items)) {
```
Comment (tractorcow) [inline src/ORM/ArrayList.php]: Probably should do this before the `->first()` since it's wasted effort.

File: src/ORM/ArrayList.php [hunk 4]
```diff
@@ -312,6 +316,10 @@ public function first()
      */
     public function last()
     {
+        if(empty($this->items)) {
```
Comment (tractorcow) [inline src/ORM/ArrayList.php]: Check linting.

File: src/ORM/ArrayList.php (hunk 3 above)
Comment (tractorcow) [inline src/ORM/ArrayList.php]: or do `if (empty($firstRecord))`

**Pull request merged**